### PR TITLE
Feat: jaqpot-223

### PR DESCRIPTION
**Pull request by johnsaveus**
Torch Graph Inference for Binary Classification

### CODE REVIEW

**Review by vassilismin — APPROVED** [review]

File: src/handlers/predict.py [hunk 1]
```diff
@@ -1,28 +1,104 @@
...
+
+def graph_post_handler(request: PredictionRequestPydantic):
+    # Obtain the request info
+    onnx_model = base64.b64decode(request.model["rawModel"])
+    ort_session = onnxruntime.InferenceSession(onnx_model)
+    feat_config = request.extraConfig["torchConfig"]["featurizer"]
+    # Load the featurizer
+    featurizer = SmilesGraphFeaturizer()
+    featurizer.load_json_rep(feat_config)
+    smiles = request.dataset["input"][0]["SMILES"]
```
Comment (alarv) [inline src/handlers/predict.py:52]: This only works for SmilesGraphFeaturizer and smiles single input as an independent feature, but what about the rest of the torch models that users may upload?

Comment (johnsaveus) [reply src/handlers/predict.py:52]: The SmilesGraphFeaturizer works for all the torch models at the moment.

Comment (alarv) [reply src/handlers/predict.py:53]: Then we should rename it to `JaqpotGraphFeaturizer`. Also we should remove hardcoding the dataset["input"][0]["SMILES"] as this assumes that the model input is always going to be an array with a single smiles inside, but that won't always be the case :/

File: src/handlers/predict.py [hunk 2]
```diff
@@ -1,28 +1,104 @@
...
```
Comment (alarv) [inline src/handlers/predict.py:52]: No need to hardcoded the smiles here, just upload the input to the model

**Review by alarv — CHANGES_REQUESTED** [review]

File: src/handlers/predict.py (hunk 2 above)
Comment (johnsaveus) [reply src/handlers/predict.py]: fixed. Need approval for merge

File: src/handlers/predict.py [hunk 3]
```diff
@@ -1,28 +1,104 @@
...
+    smiles = request.dataset["input"][0]
 
+    def to_numpy(tensor):
+        return (
+            tensor.detach().cpu().numpy()
+            if tensor.requires_grad
+            else tensor.cpu().numpy()
+        )
+
+    data = featurizer.featurize(smiles)
```
Comment (alarv) [inline src/handlers/predict.py:61]:
```suggestion
    def featurize_smiles_array(smiles_array, featurizer):
    if not isinstance(smiles_array, list):
        raise ValueError("Input must be a list of SMILES strings")

    featurized_data = []
    for smiles in smiles_array:
        try:
            features = featurizer.featurize(smiles)
            featurized_data.append(features)
        except Exception as e:
            print(f"Error featurizing SMILES: {smiles}. Error: {str(e)}")
            featurized_data.append(None)  # or handle the error as appropriate

    return featurized_data

# Usage
smiles_array = request.dataset["input"]
data = featurize_smiles_array(smiles_array, featurizer)
```

File: src/handlers/predict.py [hunk 4]
```diff
@@ -1,28 +1,104 @@
...
+
+def graph_post_handler(request: PredictionRequestPydantic):
+    # Obtain the request info
+    onnx_model = base64.b64decode(request.model["rawModel"])
+    ort_session = onnxruntime.InferenceSession(onnx_model)
+    feat_config = request.extraConfig["torchConfig"]["featurizer"]
+    # Load the featurizer
+    featurizer = SmilesGraphFeaturizer()
+    featurizer.load_json_rep(feat_config)
+    smiles = request.dataset["input"][0]
```
Comment (alarv) [inline src/handlers/predict.py:52]:
```suggestion
def featurize_smiles_array(smiles_array, featurizer):
    if not isinstance(smiles_array, list):
        raise ValueError("Input must be a list of SMILES strings")

    featurized_data = []
    for smiles in smiles_array:
        try:
            features = featurizer.featurize(smiles)
            featurized_data.append(features)
        except Exception as e:
            print(f"Error featurizing SMILES: {smiles}. Error: {str(e)}")
            featurized_data.append(None)  # or handle the error as appropriate

    return featurized_data

# Usage
smiles_array = request.dataset["input"]
data = featurize_smiles_array(smiles_array, featurizer)
```

Comment (alarv) [reply src/handlers/predict.py:52]: so this is as generic as possible and we don't need to change it in the future. Hardcoding [0] of the input means that if we ever upload 2 smiles inputs this won't work and we won't know why till we find this [0] here 😄

**Review by alarv — APPROVED** [review]
Approving so we can merge and we'll fix the comments on another PR

**Pull request merged**